### PR TITLE
feat(cawp): county-level Women in US Congress by race

### LINKED ISSUE #4813: data(cawp): preserve district numbers from unitedstates.io JSON
Part of #4812 / milestone: County-Level CAWP Data

Currently `get_us_congress_df()` in `cawp.py` groups by `state_postal + time_period`, discarding the `district` field on each House term.

- Retain `district` number per House member row before the groupby
- Senators have no district number; assign a sentinel (e.g. `"0"` or `"S"`) to distinguish them
- Update test fixtures in `python/tests/data/cawp/` to include district field in mock JSON

### LINKED ISSUE #4814: data(cawp): check in 2022 Census county-to-congressional-district crosswalk
Part of #4812 / milestone: County-Level CAWP Data

Download the 2022 Census TIGER county-to-congressional-district crosswalk file and commit it to `data/cawp/`.

- Source: Census TIGER program (2022 redistricting cycle)
- Filename: `data/cawp/natl_cocd_2022.txt` (or similar)
- Verify the file format: state FIPS, county FIPS, congressional district number, area/population overlap columns
- Add a comment in `cawp.py` documenting the source URL and year, with a note that this must be refreshed each redistricting cycle

### LINKED ISSUE #4815: data(cawp): implement county-level joins and generate county rows
Part of #4812 / milestone: County-Level CAWP Data

Depends on #4813 and #4814.

Add a `generate_county_breakdown()` function (or extend `generate_breakdown()`) in `cawp.py`:

- Load the crosswalk from `data/cawp/natl_cocd_2022.txt`
- For each House member: join district number to crosswalk to get county FIPS list
- For each Senator: attribute to all counties in their state
- Group by `county_fips + time_period + race`: compute counts, names lists, pct metrics
- Denominator per county = 2 senators + count of House districts overlapping the county
- Add a `mock_crosswalk.txt` fixture to `python/tests/data/cawp/` for unit testing

### LINKED ISSUE #4816: data(cawp): write county-level BigQuery tables
Part of #4812 / milestone: County-Level CAWP Data

Depends on the county joins issue.

Following the existing pattern in `cawp.py`:

- Generate `race_and_ethnicity_county_current` and `race_and_ethnicity_county_historical` tables using `make_bq_table_id` with `COUNTY_LEVEL`
- Call `add_df_to_bq()` for both tables inside `write_to_bq()`
- Update `test_cawp.py` to assert these two new `add_df_to_bq` calls
- Generate golden data CSVs in `python/tests/data/cawp/golden_data/` for both new tables
- Assert against golden data in the test

### PR DESCRIPTION
**Pull request by benhammondmusic**
**Preview:** 

Screenshot 2026-06-22 at 3 22 25 PM
Screenshot 2026-06-22 at 3 22 59 PM

## Summary

- Adds county-level congressional representation tables by race (`race_and_ethnicity_county_historical` and `race_and_ethnicity_county_current`)
- Checks in Census 118th Congress county-to-congressional-district crosswalk (`data/cawp/tab20_cd11820_county20_natl.txt`) and `load_county_crosswalk()` to read it
- Adds `expand_members_to_counties()` and `get_women_congress_by_county_df()` to join congressional members to counties via the crosswalk
- Preserves House district numbers through `get_us_congress_members_df()` for county-level joins
- Switches state leg denominator source from live CAWP scrape to committed `data/cawp/cawp_state_leg_{fips}.csv` files
- Fixes All-race rows missing women's names in county output
- Skips non-numeric district codes (e.g. "ZZ" for states without defined county-district boundaries)
- Includes territory counties: normalizes Census "98" at-large code to "00" to match unitedstates.io delegate records
- Fetches Congress member JSON once per `write_to_bq()` run and passes df down; eliminates duplicate HTTP requests
- Fixes member dedup to key on `(govtrack_id, year)` rather than full-dict comparison; prevents double-counting legislators who serve two overlapping terms in the same calendar year

## Test plan

- [x] All CAWP Python tests pass (`pytest python/tests/datasources/test_cawp.py`)
- [x] Golden data for `race_and_ethnicity_county_current` and `race_and_ethnicity_county_historical` regenerated and verified to include American Samoa county 60010 with correct delegate attribution (`total_us_congress_count=1.0`, `women_this_race_us_congress_count=1.0`, `pct_share_of_us_congress=100.0`)
- [x] Spot-check BigQuery on infra-test: LA County `06037` maps to 17 districts; Baldwin County AL `01003` has `congressional_districts` populated in the current table
- [x] Spot-check BigQuery on infra-test: at least one territory county (e.g. Puerto Rico FIPS `72*`) appears in county output with non-zero `total_us_congress_count`

## Infra-test results

Deploy: [run #27984888851](https://github.com/SatcherInstitute/health-equity-tracker/actions/runs/27984888851) — success

DAG pipelines:
- [x] DAG - CAWP: [run #27985359298](https://github.com/SatcherInstitute/health-equity-tracker/actions/runs/27985359298) — success

Closes #4813 Closes #4814 Closes #4815 Closes #4816

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added comprehensive county-level congressional representation tables (membership and women-in-congress metrics) for historical and current timeframes.
  * Improved US Congress member and totals construction for consistent year-by-year counts; state legislative totals now use cached data for reliability.

* **Tests**
  * Added a mock county crosswalk fixture and new unit tests to validate county-level outputs and member-generation logic.

* **Chores**
  * Normalized spellchecker configuration encoding for consistent tooling.

### CODE REVIEW

File: python/datasources/cawp.py [hunk 1]
```diff
@@ -547,31 +601,255 @@ def scaffold_df_by_year_by_state_by_race_list(race_list: List[str], first_year:
...
+    members_county = expand_members_to_counties(members_df, crosswalk_df)
+    members_county = members_county[members_county[std_col.TIME_PERIOD_COL].isin(acs_years)]
+
+    congress_totals = (
+        members_county.groupby([std_col.COUNTY_FIPS_COL, std_col.TIME_PERIOD_COL])[NAME]
+        .apply(list)
+        .reset_index()
+        .rename(columns={NAME: std_col.CONGRESS_NAMES})
+    )
+    congress_totals[std_col.CONGRESS_COUNT] = congress_totals[std_col.CONGRESS_NAMES].apply(len).astype(float)
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py:739]: Not changing - congress_totals needs the CONGRESS_NAMES column for display in the output table; .size() would drop it.

File: python/datasources/cawp.py [hunk 2]
```diff
@@ -547,31 +601,255 @@ def scaffold_df_by_year_by_state_by_race_list(race_list: List[str], first_year:
...
+    # all-races women count per county per year
+    women_all = (
+        women_county.groupby([std_col.COUNTY_FIPS_COL, std_col.TIME_PERIOD_COL])[NAME]
+        .apply(list)
+        .reset_index()
+        .rename(columns={NAME: std_col.W_ALL_RACES_CONGRESS_COUNT})
+    )
+    women_all[std_col.W_ALL_RACES_CONGRESS_COUNT] = (
+        women_all[std_col.W_ALL_RACES_CONGRESS_COUNT].apply(len).astype(float)
+    )
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py]: Already addressed - commit 98782c463 retains the names list and populates W_ALL_RACES_CONGRESS_NAMES.

File: python/datasources/cawp.py [hunk 3]
```diff
@@ -547,31 +601,255 @@ def scaffold_df_by_year_by_state_by_race_list(race_list: List[str], first_year:
...
+    # for ALL race rows, copy all-races count to this_race count
+    df = df.merge(
+        women_by_race,
+        on=[std_col.COUNTY_FIPS_COL, std_col.TIME_PERIOD_COL, RACE_ETH],
+        how="left",
+    )
+    df[std_col.W_THIS_RACE_CONGRESS_COUNT] = df[std_col.W_THIS_RACE_CONGRESS_COUNT].fillna(0)
+    df[std_col.W_THIS_RACE_CONGRESS_NAMES] = df[std_col.W_THIS_RACE_CONGRESS_NAMES].replace("", float("nan"))
+    all_mask = df[RACE_ETH] == Race.ALL.value
+    df.loc[all_mask, std_col.W_THIS_RACE_CONGRESS_COUNT] = df.loc[all_mask, std_col.W_ALL_RACES_CONGRESS_COUNT]
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py:807]: Already addressed - commit 98782c463 copies both count and names from all-races columns to this-race for All rows.

File: python/datasources/cawp.py [hunk 4]
```diff
@@ -547,31 +601,255 @@ def scaffold_df_by_year_by_state_by_race_list(race_list: List[str], first_year:
...
+    scaffold["state_fips"] = scaffold[std_col.COUNTY_FIPS_COL].str[:2]
+    scaffold = merge_utils.merge_state_ids(scaffold, keep_postal=True)
+    scaffold = merge_utils.merge_county_names(scaffold)
+
+    # --- merge totals and women counts ---
+    df = scaffold.merge(congress_totals, on=[std_col.COUNTY_FIPS_COL, std_col.TIME_PERIOD_COL], how="left")
+    df[std_col.CONGRESS_COUNT] = df[std_col.CONGRESS_COUNT].fillna(0)
+
+    df = df.merge(women_all, on=[std_col.COUNTY_FIPS_COL, std_col.TIME_PERIOD_COL], how="left")
+    df[std_col.W_ALL_RACES_CONGRESS_COUNT] = df[std_col.W_ALL_RACES_CONGRESS_COUNT].fillna(0)
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py:795]: Already addressed in commit 98782c463.

File: python/datasources/cawp.py [hunk 5]
```diff
@@ -547,31 +601,255 @@ def scaffold_df_by_year_by_state_by_race_list(race_list: List[str], first_year:
...
+    std_col.add_race_columns_from_category_id(df)
+    df = df.drop(columns=[RACE_ETH, "state_fips"])
+
+    # --- compute pct metrics ---
+    df[std_col.PCT_OF_CONGRESS] = round(df[std_col.W_THIS_RACE_CONGRESS_COUNT] / df[std_col.CONGRESS_COUNT] * 100, 1)
+    df[std_col.PCT_OF_W_CONGRESS] = round(
+        df[std_col.W_THIS_RACE_CONGRESS_COUNT] / df[std_col.W_ALL_RACES_CONGRESS_COUNT] * 100, 1
+    ).fillna(0)
+
+    df = df.drop(columns=[std_col.W_ALL_RACES_CONGRESS_COUNT])
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py]: Already addressed - commit 98782c463 drops both all-races columns after use.

File: python/datasources/cawp.py [hunk 6]
```diff
@@ -584,14 +862,43 @@ def get_us_congress_totals_df():
...
+    preserving the House district number for use in county-level joins.
+
+    Returns:
+        df with columns: ID, NAME, TYPE, state_postal, time_period, DISTRICT.
+        DISTRICT is an int for House members, None for senators."""
+    raw_historical = gcs_to_bq_util.fetch_json_from_web(US_CONGRESS_HISTORICAL_URL)
+    raw_current = gcs_to_bq_util.fetch_json_from_web(US_CONGRESS_CURRENT_URL)
+    years = get_consecutive_time_periods()
+    entries = _build_congress_member_entries([*raw_historical, *raw_current], years)
+    return pd.DataFrame.from_dict(entries)
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py:906]: Good catch - will address lru_cache in a follow-up to keep this PR focused on the county data feature.

File: python/datasources/cawp.py [hunk 7]
```diff
@@ -31,6 +31,7 @@
 Last Updated: Feb. 2024
 """
 
+import os
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py:41]: Will add with the lru_cache change in a follow-up.

File: python/datasources/cawp.py [hunk 8]
```diff
@@ -547,31 +601,255 @@ def scaffold_df_by_year_by_state_by_race_list(race_list: List[str], first_year:
...
+    members_df = merge_utils.merge_state_ids(members_df, keep_postal=True)
+
+    state_to_counties = crosswalk_df[[std_col.COUNTY_FIPS_COL, "state_fips"]].drop_duplicates()
+
+    sens = members_df[members_df[TYPE] == "sen"].copy()
+    sens_county = sens.merge(state_to_counties, on="state_fips", how="left")
+
+    reps = members_df[members_df[TYPE] == "rep"].copy()
+    reps = reps[pd.to_numeric(reps[DISTRICT], errors="coerce").notna()].copy()
+    reps["cd_geoid"] = reps["state_fips"] + reps[DISTRICT].apply(lambda d: str(int(float(d))).zfill(2))
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py:655]: Valid - will vectorize in a follow-up.

File: python/datasources/cawp.py [hunk 9]
```diff
@@ -547,31 +601,255 @@ def scaffold_df_by_year_by_state_by_race_list(race_list: List[str], first_year:
...
 
-    # iterate through each legislator
+    state_to_counties = crosswalk_df[[std_col.COUNTY_FIPS_COL, "state_fips"]].drop_duplicates()
+
+    sens = df[df[POSITION] == POSITION_LABELS[CONGRESS]["sen"]].copy()
+    sens_county = sens.merge(state_to_counties, on="state_fips", how="left")
+
+    reps = df[df[POSITION] == POSITION_LABELS[CONGRESS]["rep"]].copy()
+    reps = reps[pd.to_numeric(reps[DISTRICT], errors="coerce").notna()].copy()
+    reps["cd_geoid"] = reps["state_fips"] + reps[DISTRICT].apply(lambda d: str(int(float(d))).zfill(2))
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py:705]: Valid - will vectorize in a follow-up.

File: python/datasources/cawp.py [hunk 10]
```diff
@@ -584,14 +862,43 @@ def get_us_congress_totals_df():
...
+    years = get_consecutive_time_periods()
+
+    entries = _build_congress_member_entries([*raw_historical_congress_json, *raw_current_congress_json], years)
 
-    # convert to df
-    df = pd.DataFrame.from_dict(us_congress_totals_list_of_dict)
+    df = pd.DataFrame.from_dict(entries)
 
     # get names of all TOTAL members in lists per row
     df = df.groupby([std_col.STATE_POSTAL_COL, std_col.TIME_PERIOD_COL])[NAME].apply(list).reset_index()
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py]: Good suggestion - will refactor get_us_congress_totals_df to reuse get_us_congress_members_df in the same follow-up as the lru_cache change.

File: python/datasources/cawp.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py:655]: Not changing - this series is small (~600 members max) and the .str accessor path carries its own overhead. apply with a lambda is not meaningfully slower here and the intent is clearer.

File: python/datasources/cawp.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/cawp.py:705]: Not changing - same reasoning as above: small series, no measurable performance difference, lambda is readable.

**Review by benhammondmusic — COMMENTED** [review]
## Health Equity Tracker: Senior Review

**PR type:** Backend data pipeline
**Mission alignment:** Adds county-level congressional representation by race, a meaningful new granularity for users analyzing who actually represents their community. This is core to HET's mission: race-disaggregated representation data at the unit of geography people actually live in.

---

### Scientific Accuracy

**[IMPORTANT] Historical years use 118th Congress (2022) boundaries**
File: `python/datasources/cawp.py`, `expand_members_to_counties()` and `generate_county_breakdown()`
Issue: The crosswalk (`tab20_cd11820_county20_natl.txt`) reflects 118th Congress boundaries from the 2022 redistricting cycle, but the historical table covers 2018-2021 where actual district boundaries were the 116th/115th. A user querying "who represented Baldwin County AL in 2018" via this table will get whoever serves the 118th Congress district that currently covers Baldwin County, not the actual 2018 representative for that county. The methodology is honestly documented in `expand_members_to_counties()`'s docstring, but the impact on historical interpretation is significant.
Suggestion: (1) Surface this caveat to end users via a methodology note on the county chart (likely belongs in #4823). (2) Add a grep-able constant like `CONGRESSIONAL_BOUNDARY_VINTAGE = "118th"` so the assumption is discoverable from code. (3) Consider scoping a follow-up issue to commit prior-cycle crosswalks (115th-117th) and select by year.

**[IMPORTANT] Small-denominator counties have unstable percentages**
File: `python/datasources/cawp.py`, `generate_county_breakdown()` ~line 810
Issue: A territory county with one delegate produces `pct_share_of_us_congress` of either 0 or 100. American Samoa's verified golden data (`pct_share=100.0`) is technically correct but visually equivalent to a county with thousands of reps where a single woman happens to be one of them. No suppression threshold applied before the percent calc.
Suggestion: Decide whether `total_us_congress_count<=1` rows should be suppressed in `_current`/`_historical`, or whether the downstream UI should suppress at render time. If the latter, capture the rule in the methodology page (#4823) so the binary 0%/100% values aren't read as strong signal.

### Sensitive Language

No issues found. Position labels ("U.S. Sen.", "U.S. Rep.", "U.S. Del.", "At-Large") are factually accurate titles. No user-facing strings introduce stigmatizing or non-person-first language.

### Maintainability

**[IMPORTANT] Magic codes "98" and "00" without named constants**
File: `python/datasources/cawp.py`, `load_county_crosswalk()` lines ~617-621
Issue: The territory-delegate normalization uses bare strings `"98"` (Census's non-voting at-large code) and `"00"` (unitedstates.io's at-large code). Without names, the comment is the only thing tying the two sides together; if someone changes one without updating the other, the territory join silently breaks again.
Suggestion:
```python
CENSUS_TERRITORY_ATLARGE_CODE = "98"  # how Census codes non-voting at-large in the crosswalk
LEGISLATORS_ATLARGE_CODE = "00"        # how unitedstates.io stores at-large delegates
```

**[IMPORTANT] File-level docstring is stale**
File: `python/datasources/cawp.py`, lines 1-32
Issue: The module docstring describes state and national tables but does not mention the new `race_and_ethnicity_county_historical` / `_current` outputs. A new contributor reading the top of the file will not know county tables exist.
Suggestion: Add a brief "County tables" paragraph noting they use the Census 118th crosswalk and the temporal caveat above.

**[IMPORTANT] Duplicate rep→county expansion logic**
File: `python/datasources/cawp.py`, `expand_members_to_counties()` ~line 645 and `get_women_congress_by_county_df()` ~line 695
Issue: Both functions build `cd_geoid = state_fips + DISTRICT.zfill(2)` and merge to the crosswalk on `GEOID_CD118_20` the same way. Two call sites is the borderline case for extraction.
Suggestion: This is borderline - leave as-is for now, but if a third use case emerges (e.g., congressional districts for state-leg, or a different sourced congress list), extract `_reps_to_counties(reps_df, crosswalk_df)` then. Worth noting with a TODO referencing #4814's follow-ups.

**Patterns followed well**
The new code uses `merge_state_ids`, `merge_county_names`, `merge_yearly_pop_numbers` from existing utils; subclasses `DataSource` appropriately; uses `CAWP_RACE_GROUPS_TO_STANDARD` and `Race.ALL.value` consistently with the rest of `cawp.py`; references shared `ACS_EARLIEST_YEAR` / `ACS_CURRENT_YEAR` constants. Test coverage exercises the full `write_to_bq` pipeline with golden data for both county tables, and the territory edge case is covered with American Samoa in the mock crosswalk.

### Accessibility

Backend PR, no UI changes. Downstream a11y for the county map/chart will be addressed in #4818 and the methodology surfacing in #4823.

---

### Tooling Suggestions

**Tooling: cspell additions for new domain vocabulary**
This PR introduces several proper nouns and acronyms that cspell will flag in future PRs touching cawp.

<details>
<summary>Suggested cspell.json additions</summary>

```json
{
  "words": [
    "Aumua",
    "Radewagen",
    "GEOID",
    "AIANAPI",
    "stleg",
    "Cloudflare",
    "ZZ"
  ]
}
```
</details>

**Tooling: pylint config to allow "ZZ" as a literal constant**
The non-numeric district filter uses `"ZZ"` as a sentinel from the Census source. pylint may flag bare two-letter strings in some configurations.

<details>
<summary>Note in pylintrc if it ever flags</summary>

No change needed today, but if a future pylint upgrade flags single-string-literal magic values, consider extracting `NON_VOTING_DISTRICT_SENTINEL = "ZZ"`.
</details>

---

*Review by Claude Code (/review skill). Blocking issues require resolution before merge; important suggestions are at the author's discretion. Note: this PR's diff vs origin/main is currently empty because its commits were inadvertently bundled into #4828's squash merge; review was performed against the original cawp commits `bac2ca954..1e45f0237`.*

File: python/tests/datasources/test_cawp.py [hunk 11]
```diff
@@ -66,6 +73,33 @@ def test_get_consecutive_time_periods():
     assert default_time_periods[-1] == "2025"  # TODO: make dynamic; see GitHub #2897
 
 
+def _load_test_legislators_json(url, *_args, **_kwargs):
+    if url == US_CONGRESS_HISTORICAL_URL:
+        filename = "test_legislators-historical.json"
+    else:
+        filename = "test_legislators-current.json"
+    with open(os.path.join(TEST_DIR, filename)) as f:
+        return json.load(f)
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/tests/datasources/test_cawp.py:85]: Fixed - added elif branch for US_CONGRESS_CURRENT_URL and explicit ValueError on unknown URLs.

File: python/tests/datasources/test_cawp.py [hunk 12]
```diff
@@ -66,6 +73,33 @@ def test_get_consecutive_time_periods():
...
+    df = get_us_congress_members_df()
+
+    assert DISTRICT in df.columns
+
+    reps = df[df["type"] == "rep"]
+    sens = df[df["type"] == "sen"]
+
+    assert reps[DISTRICT].notna().all(), "all House reps should have a district number"
+    assert sens[DISTRICT].isna().all(), "all senators should have district=None"
+
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/tests/datasources/test_cawp.py:104]: Fixed - added assert not reps.empty and assert not sens.empty before the district assertions.

**Review by eriwarr — APPROVED** [review]
